machines: Allow user to supply URL of ISO file in VM creation dialog

skobyda opened the pull request:
Closes #8295

martinpitt commented on test/verify/machineslib.py:
```diff
@@ -1365,6 +1375,7 @@ class TestCreateConfig:
         VALID_DISK_IMAGE_PATH = '/var/lib/libvirt/images/example.img'
         NOVELL_MOCKUP_ISO_PATH = '/var/lib/libvirt/novell.iso'  # libvirt in ubuntu-1604 does not accept /tmp
         NOT_EXISTENT_PATH = '/tmp/not-existent.iso'
+        VALID_DIRECT_ISO_URL = 'http://download.eng.brq.redhat.com/pub/fedora/linux/releases/29/Server/x86_64/iso/Fedora-Server-dvd-x86_64-29-1.2.iso'
```
This does not work. Our tests don't have network access, and even more so, this is behind the Red Hat internal VPN. The test could use [python3 -m httpserver](https://docs.python.org/3/library/http.server.html) to serve the local CirrOS image and test against that?

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1345,6 +1345,20 @@ def installWithErrorTest(dialog):
+            server = self.machine.spawn("python3 -m http.server 8080", "httpserver")
+
+            # Test detection of ISO file in URL
+            createTest(TestMachines.VmDialog(self, "subVmTestCreate20", sourceType='url',
```
Please remove the VM name.

KKoukiou replied:
Ah this change was not yet merged, https://github.com/cockpit-project/cockpit/pull/11491

Sorry :/
Leave it as is if this one will land first.

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1345,6 +1345,20 @@ def installWithErrorTest(dialog):
+            # Test detection of ISO file in URL
+            createTest(TestMachines.VmDialog(self, "subVmTestCreate20", sourceType='url',
+                                             location=config.ISO_URL,
+                                             memory_size=900, memory_size_unit='GiB',
```
900GB is too much, it will not return an error because we don't create it, but let's not confuse ourselves.

KKoukiou replied:
No big deal since this is the only remainder issue.

martinpitt commented on pkg/machines/scripts/create_machine.sh:
```diff
@@ -53,7 +53,7 @@ if [ "$SOURCE_TYPE" = "pxe" ]; then
     if [ "$SOURCE_TYPE" = "disk_image" ]; then
         INSTALL_METHOD="--import"
-    elif [ "${SOURCE#/}" != "$SOURCE" ] && [ -f "${SOURCE}" ]; then
+    elif ( [ "${SOURCE#/}" != "$SOURCE" ] && [ -f "${SOURCE}" ] ) || ( [ "$SOURCE_TYPE" = "url" ] && [ ${SOURCE: -4} = ".iso" ] ); then
```
`${SOURCE: -4}` is a bashism, it doesn't work in e. g. dash:
```
# SOURCE="foo.iso"; echo ${SOURCE: -4}
dash: 1: Bad substitution
```
And the script is /bin/sh. Also, this needs quoting. This can be done in a POSIX shell way with `[ "${SOURCE%.iso}" != "$SOURCE" ]`.

skobyda replied:
Addressed your comment.

martinpitt commented on pkg/machines/scripts/install_machine.sh:
```diff
@@ -50,7 +50,7 @@ fi
 if [ "$SOURCE_TYPE" = "pxe" ]; then
     INSTALL_METHOD="--pxe --network $SOURCE"
-elif [ "${SOURCE#/}" != "$SOURCE" ] && [ -f "${SOURCE}" ]; then
+elif ( [ "${SOURCE#/}" != "$SOURCE" ] && [ -f "${SOURCE}" ] ) || ( [ "$SOURCE_TYPE" = "url" ] && [ ${SOURCE: -4} = ".iso" ] ); then
```
dito

martinpitt commented:
Thanks! Keeping the "requested changes" for fixing the tests.

skobyda commented:
Yes, it seems there are more things to change in tests which will failt if you run tests with sourceType='url' and start_vm=True.Our tests it seems assume we don't have connection (even through we are trying to fix that with creating that local server. 
I'm gonna look more into it.

KKoukiou commented:
@skobyda do you need any help with this?

skobyda commented:
It seems interesting thing was happening:
- It seems when you pass url of ISO file to virt install during installation phase of VM, that url is later passed to QEMU.
- It seems QEMU downloads that file using CURL.
- **However** QEMU probably uses curl with range option (curl ... "range: bytes=N-N")
(https://lists.gnu.org/archive/html/qemu-devel/2013-06/msg02661.html
https://github.com/qemu/qemu/blob/master/block/curl.c)
- Standard python http server which I was trying to use in our tests doesn't have any support for range-bytes.
- So mockup server was added, which implements this support manually.

thanks to @marusak for helping me with this.

skobyda commented:
Another things needed to be solved:
- QEMU enables support of only https (not http) driver for certain distributions to download ISO file.
see rhel-8-0 for example: http://pkgs.devel.redhat.com/cgit/rpms/qemu-kvm/tree/qemu-kvm.spec?h=stream-rhel-rhel-8.0.0#n1687
- So in order to work on rhel (and other) images, our tests needs to contain HTTPS server. However it still needs to contain range option support.

- However, iso file which we use in our tests (/var/lib/libvirt/novell.iso) is empty file. It seems empty file cannot be downloaded with range option (which QEMU uses in it's curl driver) with https, otherwise it would throw http error: `416 REQUESTED RANGE NOT SATISFIABLE`
- So I also have to generate new non-empty file (/var/lib/libvirt/novellTmp.iso, size: 1MB), which can be used in our tests.

skobyda commented:
@martinpitt @KKoukiou Tests in this PR became quite obscure in order to make them work on all distributions :D. In summary, we had to add:
- range option support to python server we ran, because qemu curl driver uses that option
- make that server https, because qemu doesn't support http driver on certain distributions (i.e. rhel)
- instead of /var/lib/libvirt/novell.iso, use some non-empty file, so it could be downloaded with range byte option.
If you have any idea, any solution which could be easier than I propose here, I'm open to suggestions.

skobyda commented:
Simplified python https server code a bit for our use.

Antique commented:
> Another things needed to be solved:
> 
> * QEMU enables support of only https (not http) driver for certain distributions to download ISO file.
>   see rhel-8-0 for example: http://pkgs.devel.redhat.com/cgit/rpms/qemu-kvm/tree/qemu-kvm.spec?h=stream-rhel-rhel-8.0.0#n1687

This link is useless to upstream contributors and users as it's internal redhat server.

skobyda commented:
It seems Ubuntu uses version of virt-install, which contains a bug and disables functionality of this PR.
I reported the bug here: https://bugs.launchpad.net/ubuntu/+source/virt-manager/+bug/1827398

KKoukiou commented:
@skobyda there are still many test failures here, thus keeping needswork, please take a look.
For example:
https://209.132.184.41:8493/logs/pull-11516-20190430-183838-87304bb0-verify-debian-testing/log.html#72
cp: cannot create regular file '/etc/pki/ca-trust/source/anchors/cert.pem': No such file or directory


https://209.132.184.41:8493/logs/pull-11516-20190430-183838-87304bb0-verify-centos-7/log.html#92
warning: ERROR    Error validating install location: Could not find an installable distribution at 'https://localhost:8000/novell.iso': The URL could not be accessed, maybe you mistyped?

KKoukiou commented:
@skobyda this failure appears in almost every distro https://209.132.184.41:8493/logs/pull-11516-20190528-173838-900abc07-verify-fedora-30/log.html#243

Also please rebase, there are some flakes that I have already fixed.

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1776,7 +1818,7 @@ def tryCreate(self, dialog):
                 else:
                     b.wait_in_text("#vm-{0}-disks-hda-device".format(name), "disk")
-            elif dialog.storage_pool == 'No Storage' and dialog.sourceType == 'file':
+            elif (dialog.storage_pool == 'No Storage' and dialog.sourceType == 'file') or (dialog.sourceType == 'url' and dialog.start_vm):
```
If startVm will not be set the cdrom will be created anyway, no?
No need to change it, I just want to understand it.

skobyda replied:
Actually no.
If startVm is true, cdrom disk is created with iso url parsed into host properties of disk elements.
If startVm is false, cdrom disk is not created, and iso url is stored in metadata of XML.

skobyda commented:
Oh I see there is only one failure on debian-testing. That failure is caused by missing package qemu-block-extra .

skobyda commented:
All I need to do for that is put the package name into TEST_PACKAGES in bots/images/scripts/debian.setup?

KKoukiou commented:
> All I need to do to add package into debian images is to put the package name into TEST_PACKAGES in bots/images/scripts/debian.setup?

Yes, and you have to update the image, but shouldn't that be dependency of either virt-install or libvirt? It's weird that people should install some qemu package manually.

skobyda commented:
@Antique can give opinion about whose responsiblity should it be to bring this dependency?

skobyda commented:
@martinpitt I heard you are a debian mantainer too. Do you think this debian-specific dependency should be dependency of cockpit, since we are offering it's functionality?
Or virt-install/libvirt, who have it as secondary functionality?

Antique commented:
If we follow the logic that the package providing that feature in it's cli/gui should depend on the `qemu-block-extra` then it's QEMU itself.

- Fedora: it is the `qemu-system-$arch` package that pulls in the `qemu-block-curl` package
- CentOS 7: there is no extra package and it's part of QEMU
- Ubuntu: `qemu-block-extra` is installed as dependency of `qemu-system-common`
- Debian: the only one where installing qemu will not install `qemu-block-extra`

Libvirt should not depend on it as there is no libvirt-driver that uses it directly.  Libvirt is only passing the data provided in XML to QEMU and doesn't care about the content or specific protocol to figure out if it's supported by hypervisor or not.

virt-install/virt-manager is not only for QEMU but it can work with other hypervisors and not all of them has to be installed, actually none of them has to be installed and it can connect to remote machine so it should not depend on specific QEMU package asi it doesn't depend on any QEMU package.

In the same way cockpit-machines should not probably depend on it as well for the same reasons as virt-install.

martinpitt commented:
It makes little sense to split out qemu-block-extra and then have qemu-system-common depend on it -- then it may just as well be bundled. Supposedly the reason for that was to allow users who don't need the extra functionality to avoid the additional dependencies like curl, rados, glusterfs, and libiscsi. In general, it's the *user*, not the *provider* of functionality which has to have the dependency. Here it becomes a bit more philosophical whether it's cockpit or libvirt -- but it seems to me that this is both the path of least resistance, as well as keeping the spirit of the split-out package:

 * Make cockpit-machines add a `Recommends: qemu-block-extra`, and add that to bots/images/scripts/debian.setup `COCKPIT_DEPS` (it's available on all Debian/Ubuntu releases). This needs image rebuilds.
 * Make sure the test works with this.
 * Make the code get along with missing curl support gracefully, i. e. show a reasonable error message. This should ideally also get a test, which `dpkg -P qemu-block-extra` and then checks the error.

skobyda commented:
I added a recommended dependency for debian.
Also added a debian test which tests a presence of error after qemu-block-extra package is removed and installation of VM with source of iso url is attempted.
Failing tests do not fail on machine's test cases (apart from ubuntu-1804, but that is disk related error and I cannot think how this PR could cause that)

martinpitt commented on test/verify/machineslib.py:
```diff
@@ -1274,6 +1274,63 @@ def installWithErrorTest(dialog):
+                                                           location=config.ISO_URL,
+                                                           memory_size=256, memory_size_unit='MiB',
+                                                           storage_size=0, storage_size_unit='MiB',
+                                                           os_vendor=config.UNSPECIFIED_VENDOR,
```
That's a little thin to ensure that the error message makes sense, but :man_shrugging: - it shouldn't happen very often in the first place. Thanks!
